### DISCUSSION TITLE
Upgrading @types/react to the latest causes an error with the auth actions

### DISCUSSION THREAD

**Discussion by mhamill2** [vercel/next.js#72778] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.0.0: Tue Sep 24 23:37:36 PDT 2024; root:xnu-11215.1.12~1/RELEASE_ARM64_T6020
  Available memory (MB): 32768
  Available CPU cores: 12
Binaries:
  Node: 22.10.0
  npm: 10.9.0
  Yarn: 1.22.21
  pnpm: 9.12.3
Relevant Packages:
  next: 15.0.3 // Latest available version is detected (15.0.3).
  eslint-config-next: 15.0.3
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which example does this report relate to?

with-supabase

### Describe the Bug

When upgrading to a newer version of `@types/react`, the formAction on the Button component no longer accepts the signUp action as a valid type. I receive the following error message:

```
Type '(formData: FormData) => Promise<{ error: string; }>' is not assignable to type 'string | ((formData: FormData) => void | Promise<void>) | undefined'.
  Type '(formData: FormData) => Promise<{ error: string; }>' is not assignable to type '(formData: FormData) => void | Promise<void>'.
    Type 'Promise<{ error: string; }>' is not assignable to type 'void | Promise<void>'.
      Type 'Promise<{ error: string; }>' is not assignable to type 'Promise<void>'.
        Type '{ error: string; }' is not assignable to type 'void'.ts(2322)
index.d.ts(3144, 9): The expected type comes from property 'formAction' which is declared here on type 'IntrinsicAttributes & ButtonProps & RefAttributes<HTMLButtonElement>'
(property) ButtonHTMLAttributes<HTMLButtonElement>.formAction?: string | ((formData: FormData) => void | Promise<void>) | undefined
```

This is happening at least in the sign-up page when passing `signUpAction` to the `formAction` on the submit button. 
<img width="985" alt="Screenshot 2024-11-12 at 8 25 51 PM" src="https://github.com/user-attachments/assets/9209a465-73f6-4104-b629-5fd9420cf6b7">

### Expected Behavior

The action function should still be a valid type when upgrading `@types/react`

### To Reproduce

1. npx create-next-app -e with-supabase
2. Upgrade @types/react to 18.3.12 and install packages
3. See that the type is no longer valid for the formAction

**Comment by samcx** [comment]
@mhamill2 Did you upgrade the other packages as well? Not having an issue.

```
{
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@radix-ui/react-checkbox": "^1.1.1",
    "@radix-ui/react-dropdown-menu": "^2.1.1",
    "@radix-ui/react-label": "^2.1.0",
    "@radix-ui/react-slot": "^1.1.0",
    "@supabase/ssr": "latest",
    "@supabase/supabase-js": "latest",
    "autoprefixer": "10.4.20",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.1",
    "geist": "^1.2.1",
    "lucide-react": "^0.456.0",
    "next": "latest",
    "next-themes": "^0.4.3",
    "prettier": "^3.3.3",
    "react": "18.3.1",
    "react-dom": "18.3.1"
  },
  "devDependencies": {
    "@types/node": "22.9.0",
    "@types/react": "18.3.12",
    "@types/react-dom": "18.3.1",
    "postcss": "8.4.49",
    "tailwind-merge": "^2.5.2",
    "tailwindcss": "3.4.14",
    "tailwindcss-animate": "^1.0.7",
    "typescript": "5.6.3"
  },
  "packageManager": "pnpm@9.12.3"
}
```

![CleanShot 2024-11-13 at 12 06 03@2x](https://github.com/user-attachments/assets/51a7f4a5-4601-43fb-9342-535528813d21)

**Reply by mhamill2** [reply]
@samcx Yeah I'm not seeing any issues when running the dev server, however I see a typescript error in the app/(auth-pages)/sign-up/page.tsx file. I copied and installed your dependencies exactly and I still see the error. 

```
# cd app/\(auth-pages\)/sign-up 
# npx tsc --noEmit 
page.tsx:42:25 - error TS2322: Type '(formData: FormData) => Promise<{ error: string; }>' is not assignable to type 'string | ((formData: FormData) => void | Promise<void>) | undefined'.
  Type '(formData: FormData) => Promise<{ error: string; }>' is not assignable to type '(formData: FormData) => void | Promise<void>'.
    Type 'Promise<{ error: string; }>' is not assignable to type 'void | Promise<void>'.
      Type 'Promise<{ error: string; }>' is not assignable to type 'Promise<void>'.
        Type '{ error: string; }' is not assignable to type 'void'.

42           <SubmitButton formAction={signUpAction} pendingText="Signing up...">
                           ~~~~~~~~~~

  ../../../node_modules/@types/react/index.d.ts:3144:9
    3144         formAction?:
                 ~~~~~~~~~~
    The expected type comes from property 'formAction' which is declared here on type 'IntrinsicAttributes & ButtonProps & RefAttributes<HTMLButtonElement> & { pendingText?: string | undefined; }'


Found 1 error in page.tsx:42
```
<img width="895" alt="Screenshot 2024-11-13 at 3 20 15 PM" src="https://github.com/user-attachments/assets/0c2c0590-285d-44ad-9b1b-a94b3c542af9">



package.json:

```json
{
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start"
  },
  "dependencies": {
    "@radix-ui/react-checkbox": "^1.1.1",
    "@radix-ui/react-dropdown-menu": "^2.1.1",
    "@radix-ui/react-label": "^2.1.0",
    "@radix-ui/react-slot": "^1.1.0",
    "@supabase/ssr": "latest",
    "@supabase/supabase-js": "latest",
    "autoprefixer": "10.4.20",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.1",
    "geist": "^1.2.1",
    "lucide-react": "^0.456.0",
    "next": "latest",
    "next-themes": "^0.4.3",
    "prettier": "^3.3.3",
    "react": "18.3.1",
    "react-dom": "18.3.1"
  },
  "devDependencies": {
    "@types/node": "22.9.0",
    "@types/react": "18.3.12",
    "@types/react-dom": "18.3.1",
    "postcss": "8.4.49",
    "tailwind-merge": "^2.5.2",
    "tailwindcss": "3.4.14",
    "tailwindcss-animate": "^1.0.7",
    "typescript": "5.6.3"
  },
  "packageManager": "pnpm@9.12.3"
}
```

**Reply by samcx** [reply]
@mhamill2 The issue is with this return → https://github.com/vercel/next.js/blob/2cd75cc78ca8e0de4dc5917048bdf16ff6b62d75/examples/with-supabase/app/actions.ts#L15.

Pushing up a fix!

**Reply by samcx** [reply]
- x-ref: https://github.com/vercel/next.js/pull/72783

**Reply by icyJoseph** [reply]
For reference, the React type did introduce some stronger inference right?

- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70197

So like, if you pass the server action directly to a form tag, then TypeScript wants to enforce a void return type, since React doesn't really consume whatever is returned... but if, for example, you used useActionState, then you could consume the return type, and then it'd be allowed to have anything else than void.

**Reply by samcx** [reply]
> For reference, the React type did introduce some stronger inference right?

I believe so!

**Comment by leojkwan** [comment]
A similar issue is still occurring in latest canary build.

```
Type '(formData: FormData) => Promise<never>' is not assignable to type 'string'.ts(2322)
index.d.ts(3144, 9): The expected type comes from property 'formAction' which is declared here on type 'IntrinsicAttributes & ButtonProps & RefAttributes<HTMLButtonElement> & { pendingText?: string | undefined; }'
```

Had to comment out all formations like sign in / sign out/ forgot password to get starter project deployed to Vercel.

**Reply by icyJoseph** [reply]
There's no issue. The types just enforce (by inference) that, if you pass a server action to the action prop of a form, then that server action ought to return nothing. 

That's because, React does nothing with that result anyway.

**Reply by leojkwan** [reply]
@icyJoseph I'm sure you're right, but regradless- the Next.js + Supabase starter template isn't building. This thread is specifically about the `with-supabase` template. You can give it a try on https://vercel.com/templates/next.js/supabase. cc @samcx maybe you have more insight here.

**Reply by leojkwan** [reply]
To clarify, it's not the exact issue @samcx fixed; its a similar issue stemming from `When upgrading to a newer version of @types/react, the formAction on the Button component no longer accepts the signUp action as a valid type`

**Reply by icyJoseph** [reply]
Not really, it's just the floating @types/react dependency (the one with `^`). 

I think this PR, on the types, https://github.com/DefinitelyTyped/DefinitelyTyped/pull/71388, will fix it. More context here, https://github.com/vercel/next.js/discussions/73581#discussioncomment-11488364 ~ in short, it was a bit of a mistake upstream on the types

Just use the pinned version, remove `^`, of @types/react@18.3.13, and @types/react-dom@18.3.1

**Reply by leojkwan** [reply]
TY for clarifying @icyJoseph. To prevent others from being blocked running a starter Next.js + Supabase project, here is the PR https://github.com/vercel/next.js/pull/73711.

**Reply by eps1lon** [reply]
Should be fixed with latest `@types/*` i.e. `@types/react@18.3.15` and `@types/react-dom@18.3.5`

**Reply by jcllobet** [reply]
thanks so much for the package fix @eps1lon. It saved my life after a good amount of time debugging and going back in circles not understanding what I was doing wrong.
